### ISSUE TITLE
Implement wctype

### ISSUE THREAD

**Issue by dlrobertson** [redox-os/relibc#32]
Implement the `wctype` following following http://pubs.opengroup.org/onlinepubs/7908799/xsh/wctype.h.html

 - Types
   - [ ] `wctype_t`
 - Functions
   - [ ] `iswalnum`
   - [ ] `iswalpha`
   - [ ] `iswcntrl`
   - [ ] `iswdigit`
   - [ ] `iswgraph`
   - [ ] `iswlower`
   - [ ] `iswprint`
   - [ ] `iswpunct`
   - [ ] `iswspace`
   - [ ] `iswupper`
   - [ ] `iswxdigit`
   - [ ] `iswctype`
   - [ ] `towctrans` (No `unimplemented` stub)
   - [ ] `towlower`
   - [ ] `towupper`
   - [ ] `wctrans` (No `unimplemented` stub)
   - [ ] `wctype`

See https://git.musl-libc.org/cgit/musl/tree/src/ctype for a good reference implementation.

**Comment by rep-nop** [comment]
I'd like to take a stab at this!

**Comment by rep-nop** [comment]
Oh wow, how did I miss that lol. Thanks!

**Comment by rep-nop** [comment]
Is there a case to make for having subtractions/additions use `overflowing_sub`/`overflowing_add` instead of just raw subtraction/addition? Something that could be leveraged esp for debug builds, and iirc should get optimized out in release anyway.

**Comment by LeonardKoenig** [comment]
I think it should be noted that MUSL basically only implements a C.UTF-8 locale, so `iswX()` only tests against that, the complete functionality of from the `locale.h`-header is completely unimplemented, similarly `wchar.h`.

@rep-nop regarding https://github.com/rep-nop/relibc/blob/1ee0f29f7e87fc34dc0f0b574057a83c0c81fbcb/src/wctype/src/lib.rs#L82, `wint_t` is "An integer type capable of storing any valid value of wchar_t or WEOF." So I'd recommend writing:

```rust
if wc == WEOF { return -1 }

wc = wc as wchar_t; // or cast to u16, whatever
// [...]
```

@dlrobertson I'm generally curious on why it was chosen to use `i16` as `wchar_t` and not the rust `char`, if we already limit ourselves to UTF-8 -- or is it planned to change that?

**Comment by Tommoa** [comment]
@dlrobertson @LeonardKoenig 
Another thing to note is that the rust `char` primitive is effectively UTF-32, and typically wchar_t is only 16 bits and used for UTF-16. It is also usually only used on Windows as almost all Unix APIs use UTF-8, while Windows uses UTF-16.

**Comment by LeonardKoenig** [comment]
@Tommoa The spec reads that `wchar_t` needs to hold any type that's supported by the OS, namely also UTF-8 which'd be 4 Bytes (or possibly up to6  in earlier revisions).

One cannot really say that it typically holds UTF-16, arguably it usually holds ASCII since on startup the locale is C and thus C-locale codepoints are used.

To support *anything* but ASCII and UTF-8 we need `wchar_t` to be properly implemented and any UNIX program dealing with such encodings need this. Actually, on real UNIX the usage of `wchar_t` is quite widespread.

If we limit us to `wchar_t` holding UTF-8 codepoints only like musl we can at least use the utility functions for UTF-8, which is at least what we need.

Of course using just `char*` is also rather widespread, but usually only when dealing with data that "happens to be string", ie. when ignoring the actual character encoding.

**Comment by Tommoa** [comment]
@LeonardKoenig I have never seen `wchar_t` used on a Unix system. Could you point me to a place that uses it?

My understanding is that Unix almost exclusively uses UTF-8, which is adequately represented by `char`. Windows almost exclusively uses UTF-16, which can only be encapsulated by a `wchar_t`.

**Comment by Tommoa** [comment]
@LeonardKoenig I see, thanks for letting me know. I tend to only work with Linux, so I almost never use `wchar_t` and just stick to `char`.

I'm not sure what the planned/current wide character type in Redox is, but would it then be best to have the `wchar_t` type abstracted out and given a platform specific type (along the lines of `#[cfg(target_os = "linux")] type wchar_t = `)?

**Comment by LeonardKoenig** [comment]
Yeah, it's usually okay if you know you know which encoding you deal with :)

Basically yes, that's mostly the point of `wchar_t` anyway:  To be dependent on the platform, and as need to support UTF-8 on Linux it has to be at least 4 Bytes big there, I suppose the same goes for Redox though.  The primary question being:  Do we want to support any other encodings for `wchar_t` too, or should it just be effectively a UTF-8 utility type & functions

**Comment by LeonardKoenig** [comment]
I need to write a BSc thesis in CS anyway and I'm playing with the thought of implementing "proper" `wchar_t` for this project, although I have some tempting other ideas too.  I don't promise anything, though.

**Comment by rep-nop** [comment]
Interesting! Well if someone knows much more about this stuff than me (which doesn't take much lol), I guess feel free to work off of whatever I have. Considering the lack of proper support that `musl` seems to have, this implementation is probably going to be a lot larger than I anticipated.
